packages/aws-cdk/lib/api/util/cloudformation.ts

```diff
@@ -86,32 +86,53 @@ async function waitFor<T>(valueProvider: () => Promise<T | null | undefined>, ti
 /**
  * Waits for a ChangeSet to be available for triggering a StackUpdate.
  *
+ * Will return a changeset that is either ready to be executed or has no changes.
+ * Will throw in other cases.
+ *
  * @param cfn       a CloudFormation client
  * @param stackName   the name of the Stack that the ChangeSet belongs to
  * @param changeSetName the name of the ChangeSet
  *
  * @returns       the CloudFormation description of the ChangeSet
  */
 // tslint:disable-next-line:max-line-length
-export async function waitForChangeSet(cfn: CloudFormation, stackName: string, changeSetName: string): Promise<CloudFormation.DescribeChangeSetOutput | undefined> {
+export async function waitForChangeSet(cfn: CloudFormation, stackName: string, changeSetName: string): Promise<CloudFormation.DescribeChangeSetOutput> {
   debug('Waiting for changeset %s on stack %s to finish creating...', changeSetName, stackName);
-  return waitFor(async () => {
+  const ret = await waitFor(async () => {
     const description = await describeChangeSet(cfn, stackName, changeSetName);
     // The following doesn't use a switch because tsc will not allow fall-through, UNLESS it is allows
     // EVERYWHERE that uses this library directly or indirectly, which is undesirable.
     if (description.Status === 'CREATE_PENDING' || description.Status === 'CREATE_IN_PROGRESS') {
       debug('Changeset %s on stack %s is still creating', changeSetName, stackName);
       return undefined;
-    } else if (description.Status === 'CREATE_COMPLETE') {
+    }
+
+    if (description.Status === 'CREATE_COMPLETE' || changeSetHasNoChanges(description)) {
       return description;
-    } else if (description.Status === 'FAILED') {
-      if (description.StatusReason && description.StatusReason.startsWith('The submitted information didn\'t contain changes.')) {
-        return description;
-      }
     }
+
     // tslint:disable-next-line:max-line-length
     throw new Error(`Failed to create ChangeSet ${changeSetName} on ${stackName}: ${description.Status || 'NO_STATUS'}, ${description.StatusReason || 'no reason provided'}`);
   });
+
+  if (!ret) {
+    throw new Error('Change set took too long to be created; aborting');
```

Comment:
Is this really a failure mode here? Like... some kind of a timeout?

Reply:
What is the question? Can this really happen in practice, can the code path be taken, ...?